[NSA-9317] Improve policy engine performance

JaredAMarsh opened the pull request:
_This change is documented in far greater detail in the associated work item as well as the linked Confluence page in the work item._

The policy engine was originally not built for scale, handling things one at a time for getting the policy results leads to duplicated work and a lack of concurrency, both of which lead to longer load times for pages needing that information when calling the `getPolicyApplicationResultsForUser` function.

This PR aims to fix that by implementing the following changes:
- `getPolicyApplicationResultsForUser` now takes in an array of service IDs and does the following:
  - Retrieves the service information concurrently: 9 requests total at a time as 3 requests are needed for the information and the services are chunked into groups of 3.
  - If any services contain policies then the user details are retrieved only once, as opposed to the previous implementation where it was retrieved once per service.
  - Retrieves the user's access to the services at the organisation concurrently in groups of 10 to try and avoid hammering our APIs by doing all services at once. Whilst I'm confident in our infrastructure this is already a vast improvement and I don't want to push things too far.
  - Returns an array of results for each service containing their ID, and then the policies applied, roles available, and whether the service is available to the user.

I also removed an unused API call where the endpoint no longer exists, and it wasn't being used in our code. **I went out of my way not to touch the actual policy condition checking** to ensure that didn't fail, and have merely improved how we grab the data for the services and access to improve loading times.

I also kept the way we clone the user object to avoid any side-effects as `JSON.parse(JSON.stringify(user))` and the reason for this is `Jest` didn't play nicely with the `structuredClone` function as it doesn't copy the prototype in the `Jest` runner, so `instanceof` returned the wrong result and I didn't want to change anything related to the policy comparison section to avoid breaking that.

JaredAMarsh committed:
Remove unused constraints API call

Cleaning up as the endpoint no longer exists.

JaredAMarsh committed:
Fix NPM dependency vulnerabilities

JaredAMarsh committed:
Rework policy application results to take array of service IDs

Reduces user/organisation requests. Uses concurrency on service info/access requests.

JaredAMarsh committed:
Update version to 4.0.0-alpha.1

AMcCartneyMTH commented on lib/index.js:
```diff
@@ -324,29 +372,59 @@ class PolicyEngine {
+      policyResults.push(
+        ...servicesWithPolicies.map((service, index) => {
+          const userClone = JSON.parse(JSON.stringify(user));
+          userClone.roles = userAccess[index];
```
Replace JSON cloning with shallow clone (cheaper): // instead of `JSON.parse(JSON.stringify(user)) const userClone = { ...user };` // shallow copy `userClone.roles = userAccess[index]; `

I believe a shallow copy will work here based on the shape of roles, but a deep clone using `structuredClone(..)` might be needed. Something like: ``const userClone = typeof structuredClone === 'function' ? structuredClone(user) : JSON.parse(JSON.stringify(user));``

JaredAMarsh replied:
I went with the deep clone as that was already present, and this object can be a couple of layers deep which might be affected down the line. `structuredClone` is a good alternative that I had planned to use but as explained above `jest` doesn't play nicely with it without changing the `typeof`/`instanceof` logic 😄

ipredoi reviewed (commented):
Changes seem good to me 👍 
I tested this in the Dev environment, and the page load time when associating a service to a user with 20 organisations is approximately 5.5 seconds.
Happy to approve this once the `-alpha.1` will be removed from the package version

JaredAMarsh committed:
Update version to 4.0.0

DEV/TEST testing passed, so we're happy for this to be marked stable.

JaredAMarsh committed:
Correct behaviour for retrieving user service access

Previous implementation always returned [], and no policies currently check roles correctly.

JaredAMarsh committed:
Update version to 4.0.1-alpha.1

An issue was identified for retrieving user roles.

JaredAMarsh committed:
Update version to 4.0.1

Regression has passed and this change is untouched by our current policies.

ipredoi reviewed (approved):
Changes seem good to me 👍
